src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs

```diff
@@ -620,15 +621,25 @@ protected internal override Task<HttpResponseMessage> SendAsync(HttpRequestMessa
                 return Task.FromCanceled<HttpResponseMessage>(cancellationToken);
             }
 
-            HttpMessageHandlerStage handler = _handler ?? SetupHandlerChain();
-
             Exception? error = ValidateAndNormalizeRequest(request);
             if (error != null)
             {
                 return Task.FromException<HttpResponseMessage>(error);
             }
 
-            return handler.SendAsync(request, cancellationToken);
+            return _handler is { } handler
+                ? handler.SendAsync(request, cancellationToken)
+                : CreateHandlerAndSendAsync(request, cancellationToken);
+
+            // SetupHandlerChain may block for a few seconds in some environments.
+            // E.g. during the first access of HttpClient.DefaultProxy - https://github.com/dotnet/runtime/issues/115301.
+            // The setup procedure is enqueued to thread pool to prevent the caller from blocking.
+            async Task<HttpResponseMessage> CreateHandlerAndSendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+            {
+                _handlerChainSetupTask ??= Task.Run(SetupHandlerChain);
```

Comment:
If you want to optimize this a bit more and also remove the chance of double-setup you could use `new Task(s => SetupHandlerChain(s), this) + CompareExchange and Task.Start (conditionally)` and you don't need the method result here (so a void Task is sufficient), can use `_handler` after the setup task finishes.

Reply:
Sounds reasonable.
Already merged though. Do you think it justifies a separate issue? I'm not certain, because repeated initialization has been deemed not important.

Reply:
I don't think it's worth the extra logic.